### PR TITLE
[CARBONDATA-4189] alter table validation issues

### PR DESCRIPTION
**Pull request by maheshrajus**
### Why is this PR needed?
1) Alter table duplicate columns check for dimensions/complex columns missed
2) Alter table properties with long strings for complex columns should not support
 
 ### What changes were proposed in this PR?
1) Changed the dimension columns list type in preparing dimensions columns [LinkedHashSet to Scala Seq] for handling the duplicate columns 
2) Added check for throwing an exception in case of long strings for complex columns
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5392/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3647/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3649/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5394/

**Comment by maheshrajus** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5402/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3658/

File: integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala [hunk 1]
```diff
@@ -812,6 +812,40 @@ test("test alter command for boolean data type with correct default measure valu
     sql("DROP TABLE t1")
   }
 
+  test("testing the duplicate add columns for complex data types") {
+    sql("drop table if exists alter_complex")
+    sql("create table alter_complex (a int, b string, arr1 array<int>) " +
+        "stored as carbondata")
+    intercept[ProcessMetaDataException] {
+      sql("alter table alter_complex add columns(arr1 array<int>)")
+    }
```
Comment (akkio-97) [inline integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala:821]: please mention which error message is to be expected

Comment (maheshrajus) [reply integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala:821]: OK, done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3665/

**Comment by akkio-97** [comment]
LGTM

File: integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala [hunk 2]
```diff
@@ -812,6 +812,56 @@ test("test alter command for boolean data type with correct default measure valu
...
   }
 
+  test("testing the duplicate add columns for complex data types") {
+    sql("drop table if exists alter_complex")
+    sql("create table alter_complex (a int, b string, arr1 array<int>) " +
+        "stored as carbondata")
+    var exception = intercept[ProcessMetaDataException] {
+      sql("alter table alter_complex add columns(arr1 array<int>)")
+    }
+    assert(exception.getMessage
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala:822]: This assertion looks duplicate for below code. Can refactor the code, by creating variable errMsg and can directly check intercept[] {}.getmessage.contains(errMsg)

File: integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala [hunk 3]
```diff
@@ -812,6 +812,56 @@ test("test alter command for boolean data type with correct default measure valu
...
+
+  test("testing the long string properties for complex columns") {
+    sql("drop table if exists alter_complex")
+    sql("create table alter_complex (a int, arr1 array<string>) " +
+        "stored as carbondata")
+    val exception = intercept[RuntimeException] {
+      sql("alter table alter_complex SET TBLPROPERTIES" +
+          "('LONG_STRING_COLUMNS'='arr1.val')")
+    }
+    assert(exception.getMessage
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/spark/carbondata/restructure/AlterTableValidationTestCase.scala:859]: Here also, can directly check  intercept[] {}.getmessage.contains(errMsg)

File: integration/spark/src/main/scala/org/apache/spark/util/AlterTableUtil.scala [hunk 4]
```diff
@@ -916,6 +916,10 @@ object AlterTableUtil {
         val errMsg = "LONG_STRING_COLUMNS column: " + col +
                      " is not a string datatype column"
         throw new MalformedCarbonCommandException(errMsg)
+      } else if (colSchemas.exists(x => x.getColumnName.equalsIgnoreCase(col) &&
```
[the remark this answers is not in the document]
Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/util/AlterTableUtil.scala:919]: ok,done

File: integration/spark/src/main/scala/org/apache/spark/util/AlterTableUtil.scala [hunk 5]
```diff
@@ -911,8 +911,12 @@ object AlterTableUtil {
                        " does not exist in table. Please check the DDL."
         throw new MalformedCarbonCommandException(errorMsg)
       } else if (colSchemas.exists(x => x.getColumnName.equalsIgnoreCase(col) &&
-                                          !x.getDataType.toString
-                                            .equalsIgnoreCase("STRING"))) {
+                                        x.isComplexColumn)) {
+        val errMsg = "Complex child column " + col + " cannot be set as LONG_STRING_COLUMNS"
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/util/AlterTableUtil.scala:915]:
```suggestion
        val errMsg = s"Complex child column $col cannot be set as LONG_STRING_COLUMNS"
```

Comment (maheshrajus) [reply integration/spark/src/main/scala/org/apache/spark/util/AlterTableUtil.scala]: done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3670/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5414/

**Comment by Indhumathi27** [comment]
LGTM

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5415/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3672/

**Pull request closed without merging**